### PR TITLE
Refactor device handling

### PR DESCRIPTION
**Pull request by amshinde**
Refactor the way devices are handled so that we store an invariable config set in the configuration.
This PR moves the Device interfaces from ContainerConfig to Container and create the device implementations at create stage.
Devices are store separately in "devices.json". This simplifies the custom marshalling and unmarshalling to an extent.

### CODE REVIEW

File: device.go [hunk 1]
```diff
@@ -172,9 +125,12 @@ func (device *VFIODevice) attach(h hypervisor) error {
 
 		device.BDF = deviceBDF
 
-		if err := h.addDevice(device, vfioDev); err != nil {
+		if err := h.addDevice(*device, vfioDev); err != nil {
```
Comment (sameo) [inline device.go]: This is a separate fix, please submit it as a separate patch.

File: container.go [hunk 2]
```diff
@@ -452,6 +356,18 @@ func newContainer(pod *Pod, contConfig ContainerConfig) (*Container, error) {
 		c.mounts = mounts
 	}
 
+	devices, err := c.fetchDevices()
+	if err == nil {
+		c.devices = devices
+	} else {
+		// If devices were not found in storage, create Device implementations
+		// from the configuration.
```
Comment (sameo) [inline container.go]: > If devices were not found in storage

Is that solely to handle the `create` to `start` transition ?
It would look better if we could have the same code logic for devices as we do for mounts, process, etc...i.e. we use the Container config information as the default one and it gets overwritten by what we found on storage.

File: device.go [hunk 3]
```diff
@@ -59,6 +59,7 @@ const (
 type Device interface {
 	attach(hypervisor) error
 	detach(hypervisor) error
+	getDeviceType() string
```
Comment (sameo) [inline device.go]: Just `type() string` please. See https://golang.org/doc/effective_go.html#Getters

Comment (amshinde) [reply device.go]: Changed this to deviceType(). golang isnt happy with type()

File: device.go (hunk 1 above)
Comment (amshinde) [reply device.go]: Done

File: container.go (hunk 2 above)
Comment (amshinde) [reply container.go]: done

File: container.go [hunk 4]
```diff
@@ -422,6 +327,11 @@ func newContainer(pod *Pod, contConfig ContainerConfig) (*Container, error) {
 		return &Container{}, fmt.Errorf("Invalid container configuration")
 	}
 
+	devices, err := newDevices(contConfig.DeviceInfos)
```
Comment (sboeuf) [inline container.go]: This is gonna be called all the time, even when we don't need it (when we can actually retrieve devices from the runtime storage). Instead I think we should call this only if we cannot retrieve anything from storage, OR the other solution is to make the difference between the first time this container has been created and all other times when it is only recreated.

Comment (amshinde) [reply container.go]: @sboeuf This is what I was doing earlier, but changed it later.
@sameo Giving it a bit more thought, I like the earlier approach of creating the devices only when fetch from storage was not successful.

File: filesystem.go [hunk 5]
```diff
@@ -48,6 +48,9 @@ const (
 
 	// mountsFileType represents a mount file type
 	mountsFileType
+
+	// devicesFileType represents a mount file type
```
[the remark this answers is not in the document]
Comment (amshinde) [reply filesystem.go]: will fix

File: filesystem.go [hunk 6]
```diff
@@ -175,6 +183,55 @@ func (fs *filesystem) storeFile(file string, data interface{}) error {
...
+			Data: tempJSON,
+		}
+		typedDevices = append(typedDevices, typedDevice)
+	}
+
+	jsonOut, err := json.Marshal(typedDevices)
+	if err != nil {
+		return fmt.Errorf("Could not marshal devices: %s", err)
+	}
+	f.Write(jsonOut)
```
[the remark this answers is not in the document]
Comment (amshinde) [reply filesystem.go]: will fix

File: container.go [hunk 7]
```diff
@@ -81,110 +80,8 @@ type ContainerConfig struct {
...
-
-					c.Devices = append(c.Devices, device)
-				}
-			}
-		}
-	}
-
-	return nil
+	// Device configuration for devices that must be available within the container.
+	DeviceInfos []DeviceInfo
```
[the remark this answers is not in the document]
Comment (amshinde) [reply container.go]: I prefer leaving this as is, we were are storing multiple DeviceInfo

[the remark this answers is not in the document]
Comment (sboeuf) [reply container.go]: fair enough

File: container.go (hunk 4 above)
Comment (sameo) [reply container.go]: @amshinde It's more efficient, I agree. We can keep it there, but please add some further comments describing why and how we can reach each side of the if switch.

File: filesystem.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (amshinde) [reply filesystem.go]: Fixed

File: filesystem.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (amshinde) [reply filesystem.go]: Fixed

File: container.go (hunk 4 above)
Comment (amshinde) [reply container.go]: Fixed

**Comment by amshinde** [comment]
@sameo @sboeuf I have pushed changes.

**Comment by sboeuf** [comment]
@sameo please merge if you're okay with this PR
LGTM
